Fix the recovery of a descriptor given a PSBT

afilini opened the pull request:
This commit upgrades `rust-miniscript` with a fix to only return the prefix that
matches a `hd_keypath` instead of the full derivation path, and then adapts the
signer code accordingly.

This commit closes #108 and #109.

afilini committed:
Fix the recovery of a descriptor given a PSBT

This commit upgrades `rust-miniscript` with a fix to only return the prefix that
matches a `hd_keypath` instead of the full derivation path, and then adapts the
signer code accordingly.

This commit closes #108 and #109.

afilini commented:
@sgeisler I tried running your `multisig.rs` example from #109 with this fix and it seemed to be working nicely. I'll give you some time to play around with this before merging, let me know if you have any problems/questions!

sgeisler commented on src/descriptor/mod.rs:
```diff
@@ -358,13 +358,12 @@ impl DescriptorMeta for Descriptor<DescriptorPublicKey> {
                             .unwrap()
                             .into_iter()
-                            .skip(prefix.len())
+                            .skip(prefix.into_iter().count())
```
We really need to fix the ergonomics `DerivationPath` :grimacing:

afilini replied:
Yeah it's kinda weird at the moment because you keep mixing up `DerivationPath` with slices or vecs of `ChildNumber`

sgeisler reviewed (approved):
Looks good!

LLFourn committed:
[keys] Less convoluted entropy generation

Since const generics aren't in rust yet you have to find some awkward
workarounds. This improves the workaround for specifying entropy length.

afilini commented:
I had to make a couple more changes because the signer needs to use the full path and not just the suffix to derive the `xprv`. While I was at it, I also added a check to make sure that when we derive an `xprv` we get the expected public key as it is specified in the `hd_keypaths`.

Also added a simple signing test using a key with a longer path that would have failed before those last changes.

afilini commented:
Rebased on #111 which I'm planning to merge first. I'm waiting for travis to start working again before I can merge, since `master` is protected and requires the CI to pass.
